### ISSUE TITLE
Typed API but returning a dynamic object

### ISSUE THREAD

**Issue by monczszilard** [object/Simple.OData.Client#469]
Our Odata interface entities may be logically divided into two parts - the standard fields and the custom fields. The standard fields are well known and stable (Id, Name, ...) while the custom fields (x1, x2, ...) are known only for server side (they are changing according to our customer needs anytime). The client application is prepared to work with these custom fields (we just need to put these into some dynamic object) - in sense to enumerate them and display the values of selected ones.
The goal is that no code change is needed in the client app when the custom fields of OData interface are changed, the changed custom fields should be just available (for further configuration) as they were dynamically retrieved. NOTE: nothing like derived classes etc. is used in OData interface, there is just only one entity type which contains both standard and custom fields together.

Our requirement is to have a syntax like:
`var myClass = client.ForDynamic<MyClassWithStandardFields>("MyODataEntity").Filter(x=> x.Name == "aaa").Find...`
i.e. to allow to use a type safe properties within the OData queries but returns a dynamic object including all the fields (unless selected otherwise). In this example MyODataEntity is the entity name in OData interface and it contains all fields from MyClassWithStandardFields + some additional custom fields.

The second requirement is to expose a proper metadata for the dynamic object to be able to enumerate all properties:
`var availableProperties = myClass.GetType().GetProperties(BindingFlags.Public|BindingFlags.Instance);`
It seems that the current metadata info is incorrect. (I know there is a possibility to use the basic syntax with dictionary, but our client app is prepared for dynamic object.)

Another idea how to solve that is to prepare some kind of API for customizable factory which handles the final object creation/mapping.

**Comment by phatcher** [comment]
@monczszilard Have a look at OData Open Types, this allows you to have standard properties on an entity but also include a Dictionary<string, object> which will hold the custom fields.

You can explicitly control which properties get serialized into the ODAta model with use of DataMember attributes.

There are some issues I have found with the implementation, mainly around inheritance but the straightforward case works fine

**Comment by monczszilard** [comment]
Thanks to point out that feature, it is very close to that what we want, but unfortunately in our case the server side is not in our hands and it will not support open types. We are looking only for a client side solution which allows us to work with known "partial" DTO definition and supports also the "unknown by DTO" custom fields. In fact the metadata contains all the fields of course, we just want to handle a part of them as unknown (and handle as dynamic) in the client.
In other words we want to use typed syntax for Simple.Odata.Client which returns a dynamic object.

**Comment by phatcher** [comment]
@monczszilard I would imagine that this would require a custom serializer to achieve since it sounds like you want to support mapping some actual OData properties into a client-side dictionary if they are not defined on your client entity.

The other approach is to use/adapt the dynamic api capabilities but of course you lose compile time checking etc

**Comment by phatcher** [comment]
@monczszilard One thing to try is to define your client model as being an open type with a dynamic properties container, looking at the ToObject method, if it can't find a matching property it puts it in the dynamic property container e.g.

```
var instance = instanceFactory();

IDictionary<string, object> dynamicProperties = null;
if (!string.IsNullOrEmpty(dynamicPropertiesContainerName))
{
    dynamicProperties = CreateDynamicPropertiesContainer(type, instance, dynamicPropertiesContainerName);
}

foreach (var item in source)
{
    var property = FindMatchingProperty(type, item);

    if (property != null && property.CanWrite && !property.IsNotMapped())
    {
        if (item.Value != null)
        {
            property.SetValue(instance, ConvertValue(property.PropertyType, item.Value), null);
        }
    }
    else
    {
        dynamicProperties?.Add(item.Key, item.Value);
    }
}
```
This may or may not work but it might be a way forward...
